File: 1691-minimum-number-of-days-to-disconnect-island/minimum-number-of-days-to-disconnect-island.java
public class Solution {
    
    public int minDays(int[][] grid) {
        // Count the number of islands in the grid
        int islands = countIslands(grid);
        
        // Step 1: Check if already disconnected
        if (islands != 1) {
            return 0;
        }
        
        // Step 2: Check if disconnecting with one move
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if (grid[i][j] == 1) {
                    grid[i][j] = 0;  // Temporarily change this land to water
                    if (countIslands(grid) != 1) {
                        return 1;
                    }
                    grid[i][j] = 1;  // Revert the change
                }
            }
        }
        
        // Step 3: If no single cell removal works, return 2
        return 2;
    }

    private int countIslands(int[][] grid) {
        boolean[][] visited = new boolean[grid.length][grid[0].length];
        int islandCount = 0;
        
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if (grid[i][j] == 1 && !visited[i][j]) {
                    islandCount++;
                    dfs(grid, visited, i, j);
                }
            }
        }
        
        return islandCount;
    }

    private void dfs(int[][] grid, boolean[][] visited, int i, int j) {
        if (i < 0 || i >= grid.length || j < 0 || j >= grid[0].length || visited[i][j] || grid[i][j] == 0) {
            return;
        }
        visited[i][j] = true;
        int[][] directions = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
        for (int[] direction : directions) {
            dfs(grid, visited, i + direction[0], j + direction[1]);
        }
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        
        int[][] grid1 = {{0, 1, 1, 0}, {0, 1, 1, 0}, {0, 0, 0, 0}};
        int[][] grid2 = {{1, 1}};
        
        System.out.println(solution.minDays(grid1));  // Output: 2
        System.out.println(solution.minDays(grid2));  // Output: 2
    }
}

File: 1691-minimum-number-of-days-to-disconnect-island/minimum-number-of-days-to-disconnect-island.py
class Solution(object):
    def minDays(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        def is_connected(grid):
            # Helper function to check if the grid is connected (i.e., only one island exists).
            def dfs(x, y):
                if x < 0 or x >= len(grid) or y < 0 or y >= len(grid[0]) or grid[x][y] == 0:
                    return
                grid[x][y] = 0
                for dx, dy in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
                    dfs(x + dx, y + dy)

            found = False
            for i in range(len(grid)):
                for j in range(len(grid[0])):
                    if grid[i][j] == 1:
                        if found:
                            return False  # Found another island, grid is disconnected
                        found = True
                        dfs(i, j)
            return found

        def count_islands():
            # Counts the number of islands in the grid.
            temp_grid = [row[:] for row in grid]
            islands = 0
            for i in range(len(temp_grid)):
                for j in range(len(temp_grid[0])):
                    if temp_grid[i][j] == 1:
                        islands += 1
                        dfs(i, j)
            return islands

        # If the grid is already disconnected
        if not is_connected([row[:] for row in grid]):
            return 0

        # Check if disconnecting the grid is possible in one day
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == 1:
                    grid[i][j] = 0
                    if not is_connected([row[:] for row in grid]):
                        return 1
                    grid[i][j] = 1

        # If not possible in one day, it will take 2 days
        return 2

# Example usage:
solution = Solution()
grid1 = [[0,1,1,0],[0,1,1,0],[0,0,0,0]]
grid2 = [[1,1]]
print(solution.minDays(grid1))  # Output: 2
print(solution.minDays(grid2))  # Output: 2

        